### PR TITLE
make property generic

### PR DESCRIPTION
**Pull request by twoertwein**
Attempt at fixing #4731.

### CODE REVIEW

**Comment by hauntsaninja** [comment]
Hmm, looks fairly disruptive. In practice type checkers end up special casing property to some extent; maybe this isn't something that typeshed should do.

File: stdlib/builtins.pyi [hunk 1]
```diff
@@ -920,22 +920,28 @@ class range(Sequence[int]):
...
         doc: str | None = ...,
     ) -> None: ...
-    def getter(self, fget: Callable[[Any], Any]) -> property: ...
-    def setter(self, fset: Callable[[Any, Any], None]) -> property: ...
-    def deleter(self, fdel: Callable[[Any], None]) -> property: ...
-    def __get__(self, obj: Any, type: type | None = ...) -> Any: ...
-    def __set__(self, obj: Any, value: Any) -> None: ...
+    def getter(self, fget: Callable[[_F], _G]) -> property[_F, _G]: ...
+    def setter(self, fset: Callable[[_F, _G], None]) -> property[_F, _G]: ...
+    def deleter(self, fdel: Callable[[_G], None]) -> property[_F, _G]: ...
```
Comment (Akuli) [inline stdlib/builtins.pyi:939]: Maybe the return type should be a subclass of `property` when called on an instance of the subclass? I haven't looked at mypy_primer errors yet, so I don't know whether this fixes them.

```suggestion
    def getter(self: _T, fget: Callable[[_F], _G]) -> _T: ...
    def setter(self: _T, fset: Callable[[_F, _G], None]) -> _T: ...
    def deleter(self: _T, fdel: Callable[[_G], None]) -> _T: ...
```

File: stdlib/builtins.pyi [hunk 2]
```diff
@@ -920,22 +920,28 @@ class range(Sequence[int]):
...
+    @overload
+    def __get__(self, obj: _F, type: type | None = ...) -> _G: ...
+    @overload
+    def __get__(self, obj: None, type: type | None = ...) -> property[_F, _G]: ...
```
Comment (Akuli) [inline stdlib/builtins.pyi:943]: Same here:

```suggestion
    def __get__(self: _T, obj: None, type: type | None = ...) -> _T: ...
```

File: stdlib/builtins.pyi (hunk 1 above)
Comment (twoertwein) [reply stdlib/builtins.pyi:939]: Thanks, that sounds good. I think mypy still complains, but it seems like an improvement :)

**Comment by twoertwein** [comment]
>  In practice type checkers end up special casing property to some extent; maybe this isn't something that typeshed should do.

If it is possible to properly annotate `property`, there are at least two potential benefits
1. Codebases that inherit from `property` or write a `property`-like class know how to annotate it 
2. type checkers will hopefully not need to special case `property` anymore so that subclasses of `property` work well with the previously specialized `property` (that does currently not work well with mypy):

```py
class Property(property):
    pass


class TestA:
    @property
    def x(self) -> int:
        return 1


class TestB(TestA):
    @Property
    def x(self) -> int:
        return 2

reveal_type(TestA.x)
reveal_type(TestB.x)
reveal_type(TestA().x)
reveal_type(TestB().x)
```
mypy
```
13: error: Signature of "x" incompatible with supertype "TestA"
17: note: Revealed type is "def (self: test.TestA) -> builtins.int"
18: note: Revealed type is "Any"
19: note: Revealed type is "builtins.int"
20: note: Revealed type is "Any"
```

pyright
```
17:13 - info: Type of "TestA.x" is "property"
18:13 - info: Type of "TestB.x" is "Property"
19:13 - info: Type of "TestA().x" is "int"
20:13 - info: Type of "TestB().x" is "int"
```

**Comment by jakebailey** [comment]
This idea rings a bell; I think had talked to @erictraut at some point about using generics to eliminate `property` special casing (but, maybe I'm misremembering).

**Comment by erictraut** [comment]
Type checkers already implement many special cases for properties. The pyright code base is unfortunately littered with these special cases. I'd love to get rid of them because they're difficult to maintain and often result in inconsistent behaviors and bugs.

Making `property` generic would go a long way toward eliminating many of these special cases, but it wouldn't eliminate all.

The big challenge with making `property` generic is that there's five pieces of information that need to be parameterized:
1. Whether there's a getter
2. The type that the getter returns
3. Whether there's a setter
4. The type that the setter accepts
5. Whether there's a deleter

I don't see how this can be done in a general way with the current type system.

I also don't see how this change can be made with the existing `property` without generating a lot of new type errors in existing code bases.

If I could design properties from scratch, I would make the following simplifications:
* Eliminate the idea of a deleter. No one uses them, and they simply add complexity.
* Introduce a separate class for read-only properties versus writable properties. The `setter` method on a read-only property would return a writable property.

These simplifications would allow for two generic classes, the first for read-only properties and the second for writable properties.

```python
from typing import Any, Callable, Generic, TypeVar

_TGet = TypeVar("_TGet")
_TSet = TypeVar("_TSet")

class property(Generic[_TGet]):
    fget: Callable[[Any], _TGet] | None

    def __init__(
        self,
        fget: Callable[[Any], _TGet] | None = ...,
        doc: str | None = ...,
    ) -> None:
        ...

    def setter(
        self, fset: Callable[[Any, _TSet], None]
    ) -> "writable_property[_TGet, _TSet]":
        ...

    def __get__(self, obj: object, type: type | None = ...) -> _TGet:
        ...

class writable_property(Generic[_TGet, _TSet]):
    fget: Callable[[Any], _TGet] | None
    fset: Callable[[Any, _TSet], None] | None

    def __init__(
        self,
        fget: Callable[[Any], _TGet] | None = ...,
        fset: Callable[[Any, _TSet], None] | None = ...,
        doc: str | None = ...,
    ) -> None:
        ...

    def __get__(self, obj: object, type: type | None = ...) -> _TGet:
        ...

    def __set__(self, obj: object, value: _TSet) -> None:
        ...

class Foo:
    @property
    def foo(self) -> int:
        ...

    @foo.setter
    def foo(self, value: int | None):
        ...

    reveal_type(foo)  # writable_property[int, int | None]
```

**Comment by twoertwein** [comment]
> I don't see how this can be done in a general way with the current type system.

That's unfortunate. Have there been proposals to extend the current typing system to be able to type `property` in future versions of python?

I assume the best solution meanwhile for projects is to avoid using `property` if it causes false positives due to the special casing and implement their own type-able properties.

**Comment by erictraut** [comment]
>  Have there been proposals to extend the current typing system to be able to type property in future versions of python?

Not that I'm aware of. The current type checkers already work around the problem for all common use cases, so I don't think people are feeling much pain here. Do you have a use case where current type checkers are falling short?

**Comment by twoertwein** [comment]
>  Do you have a use case where current type checkers are falling short?

Pandas uses a custom readonly property that also caches the wrapped value [cache_readonly](https://github.com/pandas-dev/pandas/blob/3aa2f3556a42371cf0fe0f102ab33226b0a5b7ab/pandas/_libs/properties.pyx#L10) I was trying to type it since it is used in many places where it currently [hides the wrapped type annotations.](https://github.com/pandas-dev/pandas/issues/41588) . I tried multiple attempts in this [PR](https://github.com/pandas-dev/pandas/pull/43252) but all of them result in many mypy errors which seem due to the special handling  of property.

If there is no perfect solution, adding many ignore statements is probably better than using a custom property.

**Comment by tyralla** [comment]
@twoertwein: did you check if (most of) the differences reported by mypy primer make sense? 

I am asking because I am currently trying to fix python/mypy#4125 in python/mypy#11349, which is relatively easy if one defines a property via the decorator syntax (`@property`).  However, things are more complicated when using the assignment syntax (`x = property(getter, ...)`.  If I am correct, this is because mypy uses much special casing for the decorator syntax but relies on typeshed for the assignment syntax.  Hence the different reveal results in the following example:

```python
class A:
    @property
    def x(self) -> int:
        ...
    def get_y(self) -> int:
        ...
    y = property(get_y)

a: A
reveal_type(A.x)  # note: Revealed type is "def (self: A) -> builtins.int"
reveal_type(a.x)  # note: Revealed type is "builtins.int"
reveal_type(A.y)  # note: Revealed type is "Any"
reveal_type(a.y)  # note: Revealed type is "Any"
```

If I modify typeshed's `property` as you suggested, mypy reveals the following:

```python
reveal_type(A.x)  # note: Revealed type is "def (self: A) -> builtins.int"
reveal_type(a.x)  # note: Revealed type is "builtins.int"
reveal_type(A.y)  # note: Revealed type is "builtins.property[A, builtins.int]"
reveal_type(a.y)  # note: Revealed type is ""builtins.int*"
```

So, maybe your attempt is a little disruptive for mypy at the moment, as @hauntsaninja suggested.  However, eventually, it could help eliminate some of mypy's limitations without introducing too much additional special casing in its source code.

**Comment by JelleZijlstra** [comment]
Going to close this as stale for now to keep the list of open PRs manageable. If you're still interested in seeing this through, feel free to reopen or open a new PR.

**Pull request closed without merging**